src/audio/pipeline/pipeline-graph.c

```diff
@@ -341,7 +341,7 @@ static int pipeline_comp_reset(struct comp_dev *current,
 	/* Propagate reset across pipelines only in the same direction
 	 * and with the same scheduling behavior
 	 */
-	if (!is_single_ppl && !is_same_sched && !comp_same_dir(current, dir))
+	if (!is_single_ppl && !(is_same_sched && comp_same_dir(current, dir)))
```

Comment:
I am still shocked that AEC and Demux meet the condition is_same_sched (How come?). First to clarify is that is_same_sched should not be interpreted to "two components share a scheduler". Instead, "pipelines that two components belong to (individually) refer to the same scheduled component" is more accurate. Apparently the latter statement won't be true since AEC locates on DMIC pipeline while Demux locates on Spkr pipeline. Both of them are double-ended (one end is PCM and the other is DAI) and the scheduled components should be played by individual DAIs.

Our case is like:
```
 PCMP ----> DEMUX ----> DAI(SSP)  # spkr ppl
             |
             v
 PCMC <----  AEC  <----- DAI(DMIC) # dmic ppl
```
Regarding component DEMUX and AEC, they are supposed to be (at least IMPOV)\
**NO** is_single_ppl, **NO** is_same_sched, and **NOT** is_same_dir

The previous commit https://github.com/thesofproject/sof/commit/2f5bf3cc92cb1fface1aeb25bd2149e869f274d6 was probably proposed to deal with the loopback reference case like
```
 PCMP ----> DEMUX ----> DAI(SSP)  # spkr ppl
             |
             |
 PCMC <------+                   # loopback ref ppl
```
where the capture pipeline (loopback ref) has no DAI to be driven. It will be performed as an auxiliary pipeline which refers to the scheduled component from Spkr pipeline. That is\
**NO** is_single_ppl, **YES** is_same_sched, and **NO** is_same_dir

And for such cases, the capture pipeline ought to be reset in accordance with Spkr pipeline, i.e. reset needs to propagate across two pipelines.

Reply:
OK, so I'm gathering that you think the behavior in-tree is correct, and that existing components rely on cross-pipeline resets[1] for correct operation?  The comments in the tree clearly say this isn't the case, but I believe it.  

I'll move this logic to a reset handler in the AEC component instead, the same way it works for trigger.  That will avoid messing with the pipeline code, which is, well, extremely fragile.

[1] To be clear: the bug I'm trying to fix is that when you have playback running with capture, then stop playback, the AEC component gets reset incorrectly while capture is still running.

Reply:
Though again it's worth pointing out that this behavior (cross-platform reset) does NOT happen with IPC4, and is therefore going to be breaking those legacy configurations anyway.  Maybe @lgirdwood @ranj063 @lyakh @plbossart et. al. want to chime in as to what the correct semantics should be?

Reply:
> [1] To be clear: the bug I'm trying to fix is that when you have playback running with capture, then stop playback, the AEC component gets reset incorrectly while capture is still running.

Same in my mind that AEC won't get reset when the playback pipeline stops (reset), i.e. reset is not propagated. My point is that `is_same_sched` should not be met in AEC cases. In other words, the playback pipeline and the capture one should have different scheduled component.

Reply:
> isn't it the "body" of the `if` that was failing to stop the propagation? Supposedly the `if` correctly detects different pipelines and different scheduling components, but then inside the `switch (dir)` instruction it doesn't trigger the `return` path and just continues the main flow of the function?

You are right. Although AFAIK such case does not support by now (in mainstream) which should be unreachable. Considering all possible cases it may have:
```
same_ppl?  same_sched?  same_dir?    propagate reset?    example cases
1          X            X            Y                   typical flow for a pipeline
0          0            0            N                   AEC and Smart Amp as depicted in #6882[3]
0          0            1            Y (unreachable[2])
0          1            0            Y                   Loopback Ref
0          1            1            Y [1]               Keyword Detection
```

[1] it depends as soon as multi-stream is supported (https://github.com/thesofproject/sof/pull/6131)

[2] it can be met when there are 2 DAI-driven pipelines in same direction and linked together, e.g. 1-to-N playback stream. However that is still the invalid case for multi-stream, which is proposed to deal with N-to-1 playback (mixer) or 1-to-N capture (input demuxing) stream only. @plbossart @lgirdwood for viz

[3] https://github.com/thesofproject/sof/pull/6882

Reply:
Yeah, this bikeshed is telling me not to touch the pipeline code. :)

I'll move this to the component and control it there.